src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -225,13 +228,27 @@ void Complete ()
 			}
 		}
 
+		static bool TypeHasDynamicInterfaceCastableImplementationAttribute (TypeDefinition type)
+		{
+			foreach (var ca in type.CustomAttributes) {
+				TypeDefinition caType = ca.AttributeType.Resolve ();
+				if (caType.Name == "DynamicInterfaceCastableImplementationAttribute" && caType.Namespace == "System.Runtime.InteropServices")
+					return true;
+			}
+			return false;
+		}
+
 		void InitializeType (TypeDefinition type)
 		{
 			if (type.HasNestedTypes) {
 				foreach (var nested in type.NestedTypes)
 					InitializeType (nested);
 			}
 
+			if (TypeHasDynamicInterfaceCastableImplementationAttribute (type) && type.Interfaces.Count > 0) {
+				_dynamicInterfaceCastableImplementationTypes.Add (type);
```

Comment:
Can you build the queue inside ProcessDynamicCastableImplementationInterfaces on the first run instead? It could add fewer elements.

Reply:
I don't know where I'd start building the queue from if I didn't do it here. We need to check all types that have the `[DynamicInterfaceCastableImplementation]` attribute and inherit at least one interface including ones that aren't marked anywhere else for the dynamic IDynamicInterfaceCastable scenario.

Reply:
I think you need the cache only just before you enter `ProcessDynamicCastableImplementationInterfaces` and not anywhere else. You would iterate all assembly types which are not marked there instead eagerly here as the cached content here will be larger than after marking.

Reply:
I've tried moving the cache creation. Let me know if you have any feedback on how I did it!